Fix `snippet` Cypress error

nemanjaglumac opened the pull request:
### Status
WIP (ready, but probably needs refactoring)

### What does this PR accomplish?
- It solves Cypress flake that was breaking our CI very often

### Additional notes:
- Error was happening because of autocomplete_suggestion.
- This "solution" is ugly and requires typing letter by letter, but it should solve this flake for good.
- The input field loses focus because of autocomplete, and I even caught it continue typing in the first next field that has focus (a fall-through effect that you can see in this video):
https://share.getcloudapp.com/04uOb4ZN
- This is what is probably happening with the rest of our flakes that are related to typing. The only difference being that when there is no input field or text area below the modal that lost focus, we don't see it typing. That's most likely the explanation for "cut" or completely missing words.

Commit:
Fix `snippet` Cypress error

nemanjaglumac commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
   it("should let you edit snippet", () => {
+    const snippet = {
+      name: "foo-snippet",
+      sql: "some extremely very very loooong snippet",
```
Obviously, I'm not going to leave it like this 😀
Just wanted to show that it works even for long strings. Before the final merge, I'll change it back to "foo", so that we don't waste time on each CI run.

flamber reviewed (approved):
Definitely a lot harder to read than other tests, but as long as it gets rid of the flaky test then :+1:

nemanjaglumac commented:
> Definitely a lot harder to read than other tests...

Probably because we use "foo" and "stuff" instead of SQL, so they need to be in single quotes.

Maybe I introduced the confusion trying to extract name and "sql" in the snippet object?
@paulrosenzweig could be the judge as he is the original author. If so, I can easily revert it back. Please let me know.

Anyway, the important part of the fix is that function at the top `clearAndRecursivelyTypeUsingLabel()`. I've tried all other "hacks". Nothing was working here (at least not consistently). Typing letter by letter seems to be the only way :/

paulrosenzweig commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
 import { signInAsNormalUser, restore, modal } from "__support__/cypress";
 
+function _clearAndRecursivelyTypeUsingLabel(label, string) {
+  // Cypress throws an error when trying to type an empty string
+  if (!string) {
+    return;
+  }
```
I don't think you need this. `for (const char of "")` won't run the loop body at all since there are no characters.

Also, I think it would be the right move to let Cypress error if someone tries to type `""`. IMO it's better to let the test author know there's an issue.

nemanjaglumac replied:
Good point(s).
Will remove it.

paulrosenzweig commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
 import { signInAsNormalUser, restore, modal } from "__support__/cypress";
 
+function _clearAndRecursivelyTypeUsingLabel(label, string) {
```
nitpick: I don't think this is recursive, so maybe `clearAndTypeEachCharacterUsingLabel` or `clearAndTypeIterativelyUsingLabel` instead?

nemanjaglumac replied:
You're absolutely right.
I was struggling with the naming here, tbh. I like "iteratively". Thanks.

paulrosenzweig commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
+    };
+    // Single quotes are easy to miss here, but are very important!
+    // We're not using SQL keywords or actual tables, so our string must be escaped.
+    const escapedSQL = `'${snippet.sql}'`;
```
sorry, another nitpick:
This is creating a string literal, but it isn't really "escaped". If `snippet.sql` were `"stu'ff"`, it wouldn't produce `"'stuff\'ff'"`. It doesn't need to be escaped, I was just calling out the misnomer.

The specific sql content isn't the point of the test here. I used a string literal to keep it simple. If `"'stuff'"` is weird with its nested quotes, I think you could change it to `1 + 1` or something.

paulrosenzweig replied:
`1 + 1` might be better actually because you can check that it's evaluated to `2` in the result.

nemanjaglumac replied:
It's not a nitpick - you're helping a lot with your comments 🙏
I'll definitely get rid of this and try with `1+1` for the second case.

paulrosenzweig commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
 
     // SQL editor should get updated automatically
-    cy.get("@ace").contains("select {{snippet: stuff-snippet}}");
+    cy.get("@ace").contains(`select {{snippet: ${snippet.name}}}`);
```
I have a slight preference for not using the `snippet` object and interpolating. I think it's easier to read when the test contains the values that you see browser.

My thoughts would probably be different if there were a ton of duplication, but here being direct only requires a few repeats of "stuff" or "1+1".

paulrosenzweig commented on frontend/test/metabase/scenarios/question/snippets.cy.spec.js:
```diff
@@ -1,59 +1,96 @@
       cy.findByText("Save").click();
     });
 
+    cy.wait("@update-snippet");
```
Why was this needed? I would have thought that the `contains` below would handle waiting for this to complete, but it's likely I'm missing something.

nemanjaglumac replied:
It's probably not.
Before I realized that autocomplete is messing with this, my first instinct was that it wasn't finding it because of the timing. But you're corect - `cy.get()` waits until it times out, so that alone should be enough.

paulrosenzweig reviewed (approved):
🚀  Thanks for digging in and squashing the flake

Commit:
Improve test

- remove snippet object
- rename function to more accurately reflect its behaviour
- remove exit condition from the hack function
- add simple math in the second case with altered assertion

nemanjaglumac reviewed (commented):
> 🚀 Thanks for digging in and squashing the flake

Thank you for more than helpful suggestions.
Updated everything. It should be working without a hiccup from now on :)
